File: modules/recall-native/AppGroupModule.swift
import ExpoModulesCore
import Foundation
import Security

public class AppGroupModule: Module {

  // Prevent double-registration of the Darwin observer across hot reloads
  private static var darwinObserverRegistered = false

  // MARK: - Keychain helpers

  private static let keychainService = "com.b3nny1nc.recall.auth"
  private static let keychainAccount = "supabase-session"
  private static let keychainAccessGroup = "9PWN6F3TK8.com.b3nny1nc.recall"

  private static func writeTokenToKeychain(_ jsonPayload: String) {
    print("[AppGroupModule] writeTokenToKeychain — starting, payload length=\(jsonPayload.count)")
    guard let data = jsonPayload.data(using: .utf8) else {
      print("[AppGroupModule] writeTokenToKeychain — failed to encode payload as UTF-8")
      return
    }
    // Delete any existing item first
    let deleteQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keychainService,
      kSecAttrAccount as String: keychainAccount,
      kSecAttrAccessGroup as String: keychainAccessGroup,
    ]
    let deleteStatus = SecItemDelete(deleteQuery as CFDictionary)
    print("[AppGroupModule] writeTokenToKeychain — SecItemDelete status=\(Int32(deleteStatus))")

    // Add new item
    let addQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keychainService,
      kSecAttrAccount as String: keychainAccount,
      kSecAttrAccessGroup as String: keychainAccessGroup,
      kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
      kSecValueData as String: data,
    ]
    let status = SecItemAdd(addQuery as CFDictionary, nil)
    let statusInt = Int32(status)
    if status == errSecSuccess {
      print("[AppGroupModule] writeTokenToKeychain — SecItemAdd SUCCESS, wrote \(data.count) bytes (status=\(statusInt))")
      // Verify the item is actually readable
      let verifyQuery: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: keychainService,
        kSecAttrAccount as String: keychainAccount,
        kSecAttrAccessGroup as String: keychainAccessGroup,
        kSecReturnData as String: false,
        kSecMatchLimit as String: kSecMatchLimitOne,
      ]
      let verifyStatus = SecItemCopyMatching(verifyQuery as CFDictionary, nil)
      if verifyStatus == errSecSuccess {
        print("[AppGroupModule] writeTokenToKeychain — read-back verify OK: item is accessible")
      } else {
        print("[AppGroupModule] writeTokenToKeychain — read-back verify FAILED status=\(Int32(verifyStatus)): item was written but cannot be read back. Access group may not be authorised by the provisioning profile.")
      }
    } else {
      print("[AppGroupModule] writeTokenToKeychain — SecItemAdd FAILED status=\(statusInt) (errSecDuplicateItem=-25299, errSecMissingEntitlement=-34018, errSecInteractionNotAllowed=-25308)")
    }
  }

  private static func deleteTokenFromKeychain() {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: keychainService,
      kSecAttrAccount as String: keychainAccount,
      kSecAttrAccessGroup as String: keychainAccessGroup,
    ]
    let status = SecItemDelete(query as CFDictionary)
    print("[AppGroupModule] deleteTokenFromKeychain — status: \(status)")
  }

  public func definition() -> ModuleDefinition {
    Name("AppGroupModule")

    Events("onShareCompleted")

    OnCreate {
      if #available(iOS 17.2, *) {
        // After prebuild the intent files compile into the main "Recall" target,
        // so the ObjC runtime name is "Recall.RecallShortcutsHelper".
        // Try both the qualified and unqualified names for robustness.
        let classNames = ["Recall.RecallShortcutsHelper", "RecallShortcutsHelper"]
        let sel = NSSelectorFromString("updateShortcutParameters")
        for name in classNames {
          if let helperClass = NSClassFromString(name) as? NSObject.Type,
             helperClass.responds(to: sel) {
            _ = helperClass.perform(sel)
            break
          }
        }
      }

      // Register Darwin notification observer for share-completed events
      guard !AppGroupModule.darwinObserverRegistered else { return }
      AppGroupModule.darwinObserverRegistered = true

      CFNotificationCenterAddObserver(
        CFNotificationCenterGetDarwinNotifyCenter(),
        Unmanaged.passUnretained(self).toOpaque(),
        { _, observer, _, _, _ in
          guard let observer = observer else { return }
          let module = Unmanaged<AppGroupModule>.fromOpaque(observer).takeUnretainedValue()
          DispatchQueue.main.async {
            module.sendEvent("onShareCompleted", ["timestamp": Date().timeIntervalSince1970])
          }
        },
        "com.b3nny1nc.recall.shareCompleted" as CFString,
        nil,
        .deliverImmediately
      )
      print("[AppGroupModule] Darwin observer registered for com.b3nny1nc.recall.shareCompleted")
    }

    OnDestroy {
      CFNotificationCenterRemoveObserver(
        CFNotificationCenterGetDarwinNotifyCenter(),
        Unmanaged.passUnretained(self).toOpaque(),
        CFNotificationName("com.b3nny1nc.recall.shareCompleted" as CFString),
        nil
      )
      AppGroupModule.darwinObserverRegistered = false
      print("[AppGroupModule] Darwin observer removed")
    }

    AsyncFunction("getContainerPath") { (groupId: String, promise: Promise) in
      let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId)
      let result = containerURL?.path
      print("[AppGroupModule] getContainerPath groupId=\(groupId) result=\(result ?? "nil")")
      promise.resolve(result)
    }

    AsyncFunction("verifyContainer") { (groupId: String, promise: Promise) in
      guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId) else {
        print("[AppGroupModule] verifyContainer groupId=\(groupId) — containerURL is nil")
        promise.resolve([
          "containerPath": "",
          "containerExists": false,
          "tokenFileExists": false,
          "tokenFileSize": 0,
          "tokenFileModifiedTimestamp": 0.0,
        ] as [String: Any])
        return
      }

      let containerPath = containerURL.path
      // containerURL being non-nil already proves the entitlement is provisioned;
      // check that the directory is actually reachable as a sanity guard
      var isDir: ObjCBool = false
      let containerExists = FileManager.default.fileExists(atPath: containerPath, isDirectory: &isDir) && isDir.boolValue

      let tokenURL = containerURL.appendingPathComponent("auth-token.json")
      let tokenPath = tokenURL.path
      let tokenFileExists = FileManager.default.fileExists(atPath: tokenPath)

      var tokenFileSize = 0
      var tokenFileModifiedTimestamp = 0.0
      if tokenFileExists,
         let attrs = try? FileManager.default.attributesOfItem(atPath: tokenPath) {
        tokenFileSize = (attrs[.size] as? Int) ?? 0
        tokenFileModifiedTimestamp = (attrs[.modificationDate] as? Date)?.timeIntervalSince1970 ?? 0.0
      }

      let result: [String: Any] = [
        "containerPath": containerPath,
        "containerExists": containerExists,
        "tokenFileExists": tokenFileExists,
        "tokenFileSize": tokenFileSize,
        "tokenFileModifiedTimestamp": tokenFileModifiedTimestamp,
      ]
      print("[AppGroupModule] verifyContainer groupId=\(groupId) result=\(result)")
      promise.resolve(result)
    }

    AsyncFunction("readLastShareExtensionError") { (groupId: String, promise: Promise) in
      guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId) else {
        print("[AppGroupModule] readLastShareExtensionError groupId=\(groupId) — containerURL is nil")
        promise.resolve(nil)
        return
      }

      let errorURL = containerURL.appendingPathComponent("share-ext-last-error.json")
      guard FileManager.default.fileExists(atPath: errorURL.path),
            let data = try? Data(contentsOf: errorURL),
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        print("[AppGroupModule] readLastShareExtensionError groupId=\(groupId) — file missing or unreadable")
        promise.resolve(nil)
        return
      }

      print("[AppGroupModule] readLastShareExtensionError groupId=\(groupId) result=\(json)")
      promise.resolve(json)
    }

    AsyncFunction("clearLastShareExtensionError") { (groupId: String, promise: Promise) in
      guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId) else {
        print("[AppGroupModule] clearLastShareExtensionError groupId=\(groupId) — containerURL is nil")
        promise.resolve(false)
        return
      }

      let errorURL = containerURL.appendingPathComponent("share-ext-last-error.json")
      guard FileManager.default.fileExists(atPath: errorURL.path) else {
        print("[AppGroupModule] clearLastShareExtensionError groupId=\(groupId) — file does not exist")
        promise.resolve(false)
        return
      }

      do {
        try FileManager.default.removeItem(at: errorURL)
        print("[AppGroupModule] clearLastShareExtensionError groupId=\(groupId) — deleted successfully")
        promise.resolve(true)
      } catch {
        print("[AppGroupModule] clearLastShareExtensionError groupId=\(groupId) — delete failed: \(error.localizedDescription)")
        promise.resolve(false)
      }
    }

    AsyncFunction("writeTokenFile") { (groupId: String, jsonPayload: String, promise: Promise) in
      print("[AppGroupModule] writeTokenFile — start groupId=\(groupId) payloadLength=\(jsonPayload.count)")

      guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId) else {
        print("[AppGroupModule] writeTokenFile — containerURL is nil for groupId=\(groupId), writing Keychain only")
        AppGroupModule.writeTokenToKeychain(jsonPayload)
        promise.resolve(false)
        return
      }
      print("[AppGroupModule] writeTokenFile — containerURL=\(containerURL.path)")

      let tokenURL = containerURL.appendingPathComponent("auth-token.json")
      guard let data = jsonPayload.data(using: .utf8) else {
        print("[AppGroupModule] writeTokenFile — failed to encode payload as UTF-8")
        AppGroupModule.writeTokenToKeychain(jsonPayload)
        promise.resolve(false)
        return
      }

      // ── Attempt 1: NSFileCoordinator ────────────────────────────────────────
      var coordinatorError: NSError?
      var writeSuccess = false
      let coordinator = NSFileCoordinator()
      print("[AppGroupModule] writeTokenFile — attempting NSFileCoordinator write to \(tokenURL.path)")
      coordinator.coordinate(writingItemAt: tokenURL, options: .forReplacing, error: &coordinatorError) { url in
        do {
          try data.write(to: url, options: .atomic)
          writeSuccess = true
          print("[AppGroupModule] writeTokenFile — wrote \(data.count) bytes via coordinator to \(url.path)")
        } catch {
          print("[AppGroupModule] writeTokenFile — coordinator write block threw: \(error.localizedDescription)")
        }
      }
      if let err = coordinatorError {
        print("[AppGroupModule] writeTokenFile — coordinator error: \(err.localizedDescription) (code=\(err.code))")
      }

      // ── Attempt 2: direct write fallback if coordinator failed ───────────────
      if !writeSuccess {
        print("[AppGroupModule] writeTokenFile — coordinator did not succeed, trying direct write fallback")
        do {
          try data.write(to: tokenURL, options: .atomic)
          writeSuccess = true
          print("[AppGroupModule] writeTokenFile — wrote \(data.count) bytes via direct fallback to \(tokenURL.path)")
        } catch {
          print("[AppGroupModule] writeTokenFile — direct fallback FAILED: \(error.localizedDescription)")
        }
      }

      if writeSuccess {
        print("[AppGroupModule] writeTokenFile — final outcome: SUCCESS (\(data.count) bytes)")
      } else {
        print("[AppGroupModule] writeTokenFile — final outcome: ALL WRITE PATHS FAILED")
      }

      // ── Post-write sandbox verification ─────────────────────────────────────────
      if writeSuccess {
        let verifyExists = FileManager.default.fileExists(atPath: tokenURL.path)
        let verifySize = (try? FileManager.default.attributesOfItem(atPath: tokenURL.path))?[.size] as? Int ?? 0
        if !verifyExists || verifySize == 0 {
          print("[AppGroupModule] writeTokenFile — SILENT SANDBOX FAILURE: write returned success but file not found or empty (verifyExists=\(verifyExists) verifySize=\(verifySize)). The provisioning profile likely does not authorise the App Group entitlement.")
          writeSuccess = false
        } else {
          print("[AppGroupModule] writeTokenFile — post-write verify OK: file exists, size=\(verifySize)")
        }
      }

      // Always write to Keychain regardless of file write outcome
      AppGroupModule.writeTokenToKeychain(jsonPayload)

      promise.resolve(writeSuccess)
    }

    AsyncFunction("deleteTokenFile") { (groupId: String, promise: Promise) in
      guard let containerURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupId) else {
        print("[AppGroupModule] deleteTokenFile — containerURL is nil for groupId=\(groupId)")
        AppGroupModule.deleteTokenFromKeychain()
        promise.resolve(false)
        return
      }
      let tokenURL = containerURL.appendingPathComponent("auth-token.json")
      guard FileManager.default.fileExists(atPath: tokenURL.path) else {
        print("[AppGroupModule] deleteTokenFile — file does not exist, nothing to delete")
        AppGroupModule.deleteTokenFromKeychain()
        promise.resolve(false)
        return
      }
      var coordinatorError: NSError?
      var deleteSuccess = false
      let coordinator = NSFileCoordinator()
      coordinator.coordinate(writingItemAt: tokenURL, options: .forDeleting, error: &coordinatorError) { url in
        do {
          try FileManager.default.removeItem(at: url)
          deleteSuccess = true
          print("[AppGroupModule] deleteTokenFile — deleted \(url.path)")
        } catch {
          print("[AppGroupModule] deleteTokenFile — delete failed: \(error.localizedDescription)")
        }
      }
      if let err = coordinatorError {
        print("[AppGroupModule] deleteTokenFile — coordinator error: \(err.localizedDescription)")
        promise.resolve(false)
        return
      }
      promise.resolve(deleteSuccess)
      AppGroupModule.deleteTokenFromKeychain()
    }

    AsyncFunction("verifyKeychainItem") { (promise: Promise) in
      let query: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: "com.b3nny1nc.recall.auth",
        kSecAttrAccount as String: "supabase-session",
        kSecAttrAccessGroup as String: "9PWN6F3TK8.com.b3nny1nc.recall",
        kSecReturnData as String: true,
        kSecMatchLimit as String: kSecMatchLimitOne,
      ]
      var result: AnyObject?
      let status = SecItemCopyMatching(query as CFDictionary, &result)
      let present = status == errSecSuccess
      let dataSize = (result as? Data)?.count ?? 0
      print("[AppGroupModule] verifyKeychainItem — present=\(present) dataSize=\(dataSize) status=\(Int32(status))")
      promise.resolve(["present": present, "dataSize": dataSize] as [String: Any])
    }
  }
}

File: modules/recall-native/src/AppGroupModule.ts
// recall-native AppGroupModule
import { requireNativeModule } from 'expo-modules-core';
import { Platform } from 'react-native';

const APP_GROUP_ID = 'group.com.b3nny1nc.recall';

let _module: any = null;
let _moduleLoadAttempted = false;
let _moduleLoadError: string | null = null;

function getNativeModule() {
  if (_moduleLoadAttempted) return _module;
  _moduleLoadAttempted = true;
  try {
    _module = requireNativeModule('AppGroupModule');
    console.log('[AppGroupModule] requireNativeModule succeeded — module loaded');
  } catch (e: any) {
    _module = null;
    _moduleLoadError = String(e);
    console.warn(
      '[AppGroupModule] requireNativeModule FAILED:',
      _moduleLoadError,
      '| isNativeModuleError:',
      e?.code ?? 'no code',
      '| type:',
      Object.prototype.toString.call(e)
    );
  }
  return _module;
}

/**
 * Get the real App Group container path on iOS.
 * Returns null on Android or if the module is unavailable.
 */
export async function getAppGroupContainerPath(): Promise<string | null> {
  if (Platform.OS !== 'ios') return null;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] getAppGroupContainerPath — native module unavailable, returning null');
    return null;
  }
  try {
    const result = await mod.getContainerPath(APP_GROUP_ID);
    console.log('[AppGroupModule] getAppGroupContainerPath result:', result);
    return result;
  } catch (e: any) {
    console.error('[AppGroupModule] getContainerPath threw:', String(e));
    return null;
  }
}

export interface ContainerVerification {
  containerPath: string;
  containerExists: boolean;
  tokenFileExists: boolean;
  tokenFileSize: number;
  tokenFileModifiedTimestamp: number;
}

/**
 * Calls the native verifyContainer function to inspect the App Group directory
 * and the auth-token.json file. Returns null on Android or if unavailable.
 */
export async function verifyAppGroupContainer(): Promise<ContainerVerification | null> {
  if (Platform.OS !== 'ios') return null;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] verifyAppGroupContainer — native module unavailable');
    return null;
  }
  try {
    const result = await mod.verifyContainer(APP_GROUP_ID);
    console.log('[AppGroupModule] verifyContainer result:', JSON.stringify(result));
    return result ?? null;
  } catch (e: any) {
    console.error('[AppGroupModule] verifyContainer threw:', String(e));
    return null;
  }
}

export interface AppGroupDiagnostics {
  platform: string;
  moduleAvailable: boolean;
  containerPath: string | null;
  error: string | null;
  moduleLoadError: string | null;
  containerExists: boolean | null;
  tokenFileExists: boolean | null;
  tokenFileSize: number | null;
  tokenFileModifiedTimestamp: number | null;
}

/**
 * Run a full diagnostic of the App Group bridge and return a single object
 * suitable for logging. Safe to call from any platform — non-iOS fields will
 * be null.
 */
export async function getDiagnostics(): Promise<AppGroupDiagnostics> {
  const platform = Platform.OS;
  let moduleAvailable = false;
  let containerPath: string | null = null;
  let error: string | null = null;
  let verification: ContainerVerification | null = null;

  if (platform === 'ios') {
    const mod = getNativeModule();
    moduleAvailable = mod !== null;

    if (mod) {
      try {
        containerPath = await mod.getContainerPath(APP_GROUP_ID);
      } catch (e: any) {
        error = `getContainerPath threw: ${String(e)}`;
        console.error('[AppGroupModule] getDiagnostics — getContainerPath error:', error);
      }

      try {
        verification = await mod.verifyContainer(APP_GROUP_ID);
      } catch (e: any) {
        const verifyErr = `verifyContainer threw: ${String(e)}`;
        error = error ? `${error}; ${verifyErr}` : verifyErr;
        console.error('[AppGroupModule] getDiagnostics — verifyContainer error:', verifyErr);
      }
    } else {
      // Use the captured load error if available, fall back to generic message
      error = _moduleLoadError ?? 'Native module AppGroupModule is not available in this build';
    }
  }

  return {
    platform,
    moduleAvailable,
    containerPath,
    error,
    moduleLoadError: _moduleLoadError,
    containerExists: verification?.containerExists ?? null,
    tokenFileExists: verification?.tokenFileExists ?? null,
    tokenFileSize: verification?.tokenFileSize ?? null,
    tokenFileModifiedTimestamp: verification?.tokenFileModifiedTimestamp ?? null,
  };
}

/**
 * Read the last share extension error/success record from the App Group.
 * Written by ShareViewController whenever loadAuthToken() fails or succeeds.
 * Returns null on Android, if the module is unavailable, or if no record exists yet.
 */
export async function readLastShareExtensionError(): Promise<Record<string, any> | null> {
  if (Platform.OS !== 'ios') return null;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] readLastShareExtensionError — native module unavailable');
    return null;
  }
  try {
    const result = await mod.readLastShareExtensionError(APP_GROUP_ID);
    console.log('[AppGroupModule] readLastShareExtensionError result:', JSON.stringify(result));
    return result ?? null;
  } catch (e: any) {
    console.warn('[AppGroupModule] readLastShareExtensionError threw:', String(e));
    return null;
  }
}

/**
 * Delete the last share extension error record from the App Group.
 * Returns true if the file was removed, false otherwise.
 */
export async function clearLastShareExtensionError(): Promise<boolean> {
  if (Platform.OS !== 'ios') return false;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] clearLastShareExtensionError — native module unavailable');
    return false;
  }
  try {
    const result = await mod.clearLastShareExtensionError(APP_GROUP_ID);
    console.log('[AppGroupModule] clearLastShareExtensionError result:', result);
    return result ?? false;
  } catch (e: any) {
    console.warn('[AppGroupModule] clearLastShareExtensionError threw:', String(e));
    return false;
  }
}

/**
 * Write the auth token JSON string directly to the App Group container
 * using native FileManager — bypasses expo-file-system's sandbox.
 * Returns true on success.
 */
export async function writeTokenFile(jsonPayload: string): Promise<boolean> {
  if (Platform.OS !== 'ios') return false;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] writeTokenFile — native module unavailable');
    return false;
  }
  try {
    const result = await mod.writeTokenFile(APP_GROUP_ID, jsonPayload);
    console.log('[AppGroupModule] writeTokenFile result:', result);
    return result ?? false;
  } catch (e: any) {
    console.error('[AppGroupModule] writeTokenFile threw:', String(e));
    return false;
  }
}

/**
 * Delete the auth token file from the App Group container using native FileManager.
 * Returns true if the file was removed, false if it didn't exist or on error.
 */
export async function deleteTokenFile(): Promise<boolean> {
  if (Platform.OS !== 'ios') return false;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] deleteTokenFile — native module unavailable');
    return false;
  }
  try {
    const result = await mod.deleteTokenFile(APP_GROUP_ID);
    console.log('[AppGroupModule] deleteTokenFile result:', result);
    return result ?? false;
  } catch (e: any) {
    console.error('[AppGroupModule] deleteTokenFile threw:', String(e));
    return false;
  }
}

/**
 * Verify whether a Keychain item exists for the auth token.
 * Returns { present: boolean, dataSize: number } or null if unavailable.
 */
export async function verifyKeychainItem(): Promise<{ present: boolean; dataSize: number } | null> {
  if (Platform.OS !== 'ios') return null;
  const mod = getNativeModule();
  if (!mod) {
    console.warn('[AppGroupModule] verifyKeychainItem — native module unavailable');
    return null;
  }
  try {
    const result = await mod.verifyKeychainItem();
    console.log('[AppGroupModule] verifyKeychainItem result:', JSON.stringify(result));
    return result ?? null;
  } catch (e: any) {
    console.error('[AppGroupModule] verifyKeychainItem threw:', String(e));
    return null;
  }
}
